_posts/2018-11-01-Inter-service-communication.md

```diff
@@ -0,0 +1,327 @@
+This becomes very useful in situations where every request needs this extra information.
+To add an interceptor, we just add an extra method that returns the OpenFeign interceptor. 
+
+{% highlight java %}
+ private RequestInterceptor requestInterceptor() {
+        return requestTemplate -> {
+            requestTemplate.header("user", username);
+            requestTemplate.header("password", password);
+            requestTemplate.header("Accept", ContentType.APPLICATION_JSON.getMimeType());
+        };
+    }
+{% endhighlight %}
+
+To enable the customization, we add the interceptor to the builder. 
+
+{% highlight java %}
+
+   @Bean
+   AuthClient authClient() {
+        return Feign.builder()
+                .requestInterceptor(requestInterceptor())
+                .target(AuthClient.class, baseServerUrl);
+    }
+{% endhighlight %}
+
+# Securing your API
+When we want to add the security layer between our services, there are a couple solutions to look at. 
+Here are a few that can be handled by OpenFeign. 
+
+## Basic
+When you want to send basic credentials you can just add an [interceptor](#intercepting-requests) for the OpenFeign client and add the username and password.
+
+## Bearer
+For only Bearer token communication, you can just pass it down in the request header of your method call. 
+
+{% highlight java %}
+//Spring
+    @Override
+    public boolean isTokenValid(@RequestHeader("Authorization") String token);
+
+//OpenFeign
+    @RequestLine("GET /auth")
+    @Headers({"Authorization: {token}", "Accept: application/hal+json"})
+    boolean isValid(@Param("token") String token);
+{% endhighlight %}
+
+
+## OAuth2
+This link provides a good explanation about the use of OAuth2 with OpenFeign. 
+[OAuth 2 interceptor](https://jmnarloch.wordpress.com/2015/10/14/spring-cloud-feign-oauth2-authentication/ ){:target="_blank"}
+
+# Creating a SOAP client
+Besides JSON encoders and decoders, you can also enable support for XML.
+If you ever have to integrate with SOAP third party APIs, OpenFeign supports it.
+There is a very detailed explanation on how to use it in the [documentation](https://github.com/OpenFeign/feign#jaxb){:target="_blank"} of OpenFeign.
+
+
+# Handling errors with the error decoder
+The OpenFeign API provides an ErrorDecoder to handle erroneous responses from servers.
+Since there are many kind of errors we can get, we want a place where we can handle each one of them accordingly. 
+An OpenFeign ErrorDecoder must be added to the configuration of the client object as you can see in the code below.
+
+{% highlight java %}
+
+   @Bean
+   MyClient myClient() {
+       return Feign.builder()
+               .errorDecoder(errorDecoder())
+               .target(MyClient.class, <url>);
+   }
+{% endhighlight %}
+
+Rather than throwing an exception in the decode method of the ErrorDecoder, you return an exception to Feign and Feign will throw it for you.
```

Comment:
```suggestion
Rather than throwing an exception in the `decode` method of the `ErrorDecoder`, you return an exception to Feign and Feign will throw it for you.
```